{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1b3ccce9-9206-4682-8598-807cf39d9f58",
   "metadata": {},
   "source": [
    "### This notebook first labels clusters of protease inhibitors from ProteinCartography run\n",
    "Full outputs from the final cartography run (carto_run3) on Zenodo:  tick_PUFs_PIs_1200_plus_toxinDB_carto_run3.zip at https://zenodo.org/records/15186244\n",
    "\n",
    "\n",
    "#### Input:\n",
    "Umap clustering (copied over from full PC run): ../outputs/carto_run/carto_output_run3/clusteringresults/carto_run3_aggregated_features_pca_umap.tsv \\\n",
    "Leiden similarity (copied over from full PC run): ../outputs/carto_run/carto_output_run3/clusteringresults/carto_run3_leiden_similarity.tsv \\\n",
    "Structural similarity (copied over from full PC run): ../outputs/carto_output_run3/carto_run/clusteringresults/carto_run3_strucluster_similarity.tsv \n",
    "\n",
    "\n",
    "#### Output:\n",
    "Upated umap with protease inhibitor cluster names added: ../datasheets/carto_run3_aggregated_features_pca_umap_updated.tsv \\\n",
    "High confidence protease inhibitor orthogroups: ../datasheets/PI_orthogroups_high_qual_06112023.tsv \\\n",
    "Low confidence protease inhibitor orthogroups: ../datasheets/PI_orthogroups_low_qual_06112023.tsv \n",
    "\n",
    "\n",
    "Ultimately, I found 7 clusters that are high quality protease inhibitors (2.5k proteins) and 4 clusters that are interesting but are much lower confidence (1.8 k proteins).  \n",
    "\n",
    "### Then, this notebook uses the structural clusters to find orthogroups of protease inhibitors to move forward for trait mapping\n",
    "\n",
    "#### Input:\n",
    "\n",
    "Orthogroup data: Sourced from [Comparative phylogenomic analysis of chelicerates points to gene families associated with long-term suppression of host detection](https://dx.doi.org/10.57844/arcadia-4e3b-bbea)\\\n",
    "Same copy of orthogroup data can also be found here: ../datasheets/Orthogroups.tsv\n",
    "\n",
    "#### Output:\n",
    "\n",
    "Orthogroups with high confidence (HQ) PI calls: ../datasheets/PI_orthogroups_high_qual_06112023.tsv \\\n",
    "Orthogroups with low confidence (LQ) PI calls: ../datasheets/PI_orthogroups_low_qual_06112023.tsv\n",
    "\n",
    "This analysis results in 14 orthogroups of high-confidence protease inhibitors, and 22 orthogroups of low confidence protease inhibitors. These 36 orthogroups were moved forward to phylogenetic profiling to associate them with the host detection suppression trait.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6942d9cc-165e-4ba0-a7e4-f2d61b0236a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sb\n",
    "import numpy as np\n",
    "import glob\n",
    "from collections import defaultdict\n",
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34a47c44-0055-4a4c-a167-d301e8514655",
   "metadata": {},
   "source": [
    "## Reading in umap clustering of structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7f20815a-9b7f-4036-a51b-dbbe083db783",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/rb/9psz38xj6h7685fqhq0kwtc40000gn/T/ipykernel_31380/175374996.py:2: DtypeWarning: Columns (9,10,12,15,18,19) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  umap = pd.read_csv(\"../outputs/carto_run/carto_output_run3/clusteringresults/carto_run3_aggregated_features_pca_umap.tsv\", sep = \"\\t\")\n"
     ]
    }
   ],
   "source": [
    "#umap clustering from cartography\n",
    "umap = pd.read_csv(\"../outputs/carto_run/carto_output_run3/clusteringresults/carto_run3_aggregated_features_pca_umap.tsv\", sep = \"\\t\")\n",
    "umap = umap.loc[umap[\"Sequence\"].notna()]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "536d4560-a6e1-4d77-a496-4f4c41294fd1",
   "metadata": {},
   "source": [
    "### Reading in cluster self-similarity TM scores to see how well the structures were clustered in this analysis\n",
    "0.2 TM is the minimum TM score that could be considered a non-random match, so we will be highly suspicious of clusters below this threshold. Also should treat clusters with 0.2-0.5 with scrutiny, above 0.5 is generally recgonized to be the same whereas lower is less certain. ProteinCartography provided two different analyses of cluster similarity - Leiden similarity (self similarity of Leiden Clusters made in this analysis) as well as Structural Cluster similarity (Foldseek default clustering). I'm going to read in both metrics, but in downstream analyses I end up using Leiden clusters , as there are far fewer of them than the Foldseek \"Structural clusters\" and they appear to represent the different major classes of proteins well.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f5712d08-11a3-4b68-a458-a27154ed13e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in self similarity - 0.2 TM is the minimum TM score that could be considered a non-random match, so we will want to remove clusters with a self similarity below this\n",
    "#also shoudl treat clusters with 0.2-0.5 with scrutiny, above 0.5 is generally recgonized to be the same wherease lower is sketchier \n",
    "LC_self_similarity = pd.read_csv(\"../outputs/carto_run/carto_output_run3/clusteringresults/carto_run3_leiden_similarity.tsv\", sep = \"\\t\",index_col=0)\n",
    "ST_self_similarity = pd.read_csv(\"../outputs/carto_run/carto_output_run3/clusteringresults/carto_run3_strucluster_similarity.tsv\", sep = \"\\t\",index_col=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ad6e79fe-543e-4f12-a1fb-9545da227038",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add self sim information to the big umap dataframe \n",
    "sim_dict = defaultdict(list)\n",
    "for cluster in ST_self_similarity.index.values.tolist():\n",
    "    TM = ST_self_similarity.loc[[cluster],[cluster]].values[0][0]\n",
    "    sim_dict[\"StruCluster\"].append(cluster)\n",
    "    sim_dict[\"within_ST_TM\"].append(TM)\n",
    "    \n",
    "ST_sim_df = pd.DataFrame(sim_dict)\n",
    "umap = umap.merge(ST_sim_df, on = \"StruCluster\", how = \"left\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27d79677-b2c8-4f5a-8f4b-3abe58d12106",
   "metadata": {},
   "source": [
    "#### Next, I explored ProteinCartography visualization and semantic analyses of the Leiden Clusters. Doing this, I identified 7 high confidence protease inhibitor clusters. High confidence means that they have a within-cluster TM score >0.2 and annotations consistent with protease inhibition function. \n",
    "\n",
    "- LC06: Kunitz domain protease inhibitors\n",
    "- LC13: Kunitz domain protease inhibitors\n",
    "- LC14: Serpin domain protease inhibitors\n",
    "- LC15: Cysteine-rich secretory proteins (CRISPs), related to some protease inhibitors\n",
    "- LC20: Serpin domain protease inhibitors\n",
    "- LC22: Cystatin protease inhibitors\n",
    "- LC29: Trypsin-inhibitor-like (TIL) protease inhibitors\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fa27e7f1-e817-4e1e-baef-69b2e3d112a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#adding shorthand names to the dataframe so I can keep track of which cluster is which\n",
    "name_dict = {\"LC06\": 'Kunitz-2',\n",
    "             \"LC13\": 'Kunitz-1',\n",
    "             \"LC14\": 'Serpin-1',\n",
    "             \"LC15\": 'CRISP-PIs',\n",
    "             \"LC20\": 'Serpin-2',\n",
    "             \"LC22\": 'Cystatin',\n",
    "             \"LC29\": \"TIL\"}\n",
    "umap[\"PI_cluster\"] = umap['LeidenCluster'].map(name_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e03c0285-3f38-46cd-8edc-177faeec13cb",
   "metadata": {},
   "source": [
    "#### Beyone the 7 \"high confidence\" protease inhibitor clusters, I also identified 3 \"lower confidence\" clusters. These clusters either had low within-cluster TM scores or weak annotations\n",
    "\n",
    "- LC02: TM score < 0.2, Kazal and insulin growth factor (IGF) annotations\n",
    "- LC05: TM score < 0.2, Kazal and Alpha-2 macroglobulin (A2M) annotations\n",
    "- LC17: TM score > 0.2, weak annotations related to Kazal domains\n",
    "- LC21: TM score < 0.2, Kazal and Whey Acidic Protein (WAP) annotations\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "aa41f58d-c5b9-4328-909d-984adbae3f92",
   "metadata": {},
   "outputs": [],
   "source": [
    "#dictionary with sketchy clusters and shorthand names\n",
    "sketch_dict = {\"LC02\": 'BadTM-Kazal-IGF',\n",
    "             \"LC05\": 'BadTM-Kazal-A2M',\n",
    "             \"LC17\": 'WeakAnnotation-Kazal',\n",
    "             \"LC21\": 'BadTM-Kazal-WAP'}\n",
    "\n",
    "umap[\"Low_Quality_PI_cluster\"] = umap['LeidenCluster'].map(sketch_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "00348569-0e5c-415b-9814-35973e912514",
   "metadata": {},
   "outputs": [],
   "source": [
    "#adding this cluster name info to the umap csv, putting in datasheets folder\n",
    "umap.to_csv(\"../datasheets/carto_run3_aggregated_features_pca_umap_updated.tsv\", sep = \"\\t\", index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40435293-e729-4be1-9b0d-49f57133ce7b",
   "metadata": {},
   "source": [
    "### The next step is to use our NovelTree analysis to identify the gene families (orthogroups) that underlie these structural clusters. \n",
    "This will enable us perform trait mapping to see which ones are postively associated with host detection-supression trait"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38728b8e-7c66-4bca-b02a-d7f11f80e9c5",
   "metadata": {},
   "source": [
    "### Reading in full set of chelicerate proteins that went into NovelTree "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9aea16ee-c667-497b-9c20-3fb7727b91cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#reading in all annoations for the full chelicerate protein dataset (novel tree input) \n",
    "all_proteins = pd.read_csv(\"../../25aacutoff_fullset_100223_annotated/all_chelicerate_proteins_annotated.csv\")\n",
    "# reformatting Organism to match with NovelTree\n",
    "all_proteins[\"Organism\"] = all_proteins[\"species_name\"].apply(lambda x: x.replace(\"-\",\" \"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f15f1c4f-2345-4924-9e36-02a3e94b4369",
   "metadata": {},
   "source": [
    "### Reading in and reshaping orthogroup table from NovelTree\n",
    "To intersect with my analyses,  I needed to reshape the orthogroups dataframe the dataframe has 1 column with gene names and the other column with orthogroup number "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ff371df0-ae19-4b8b-8c94-d4a46a6e97d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#from NovelTree run - see https://research.arcadiascience.com/pub/result-chelicerate-detection-suppression/release/2\n",
    "orthogroups = pd.read_csv(\"../datasheets/Orthogroups.tsv\", sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9824f7cf-2bf0-4edf-a780-4534192f0e54",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use melt to reshape the DataFrame\n",
    "melted_df = orthogroups.melt(id_vars='Orthogroup', var_name='Organism', value_name='gene_name')\n",
    "\n",
    "# Drop the 'Organism' column\n",
    "reshaped_df = melted_df.drop(columns='Organism')\n",
    "reshaped_df = reshaped_df.dropna()\n",
    "# Create a copy of reshaped_df\n",
    "expanded_df = reshaped_df.copy()\n",
    "\n",
    "# Split the 'Gene Name' column on commas\n",
    "expanded_df['gene_name'] = expanded_df['gene_name'].str.split(', ')\n",
    "\n",
    "# Use explode to create a new row for each gene name\n",
    "expanded_df = expanded_df.explode('gene_name')\n",
    "expanded_df.reset_index( drop = True, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd6fed03-691d-435b-b42d-bf7e435e85c6",
   "metadata": {},
   "source": [
    "### Merging orthogroup information with full list of Chelicerate proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7b55f385-776f-4946-8d55-2188a115d432",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_proteins = all_proteins.merge(expanded_df, on = \"gene_name\", how = 'outer', indicator = True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e601d613-bbd0-4562-9658-ee0bc5bac370",
   "metadata": {},
   "source": [
    "### Adding in the structural cluster information  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ef910165-8350-4942-9c1a-01430ce93bac",
   "metadata": {},
   "outputs": [],
   "source": [
    "#just grabbing info that is unique to this df so i can intersect it back with the larger chelicerate dataset \n",
    "umap_small = umap[[\"protid\", \"PI_or_PUF\",\"StruCluster\", \"LeidenCluster\", \"within_LC_TM\", \"PI_cluster\", \"Low_Quality_PI_cluster\"]]\n",
    "#adding in umap information from cartography \n",
    "all_proteins_merged = all_proteins.merge(umap_small, left_on = \"gene_name\", right_on = \"protid\", how = \"left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "16cbb93b-5768-438d-9863-840769950838",
   "metadata": {},
   "outputs": [],
   "source": [
    "#writing out large file of all chelicerate proteins (too big for github)\n",
    "all_proteins.to_csv(\"../../25aacutoff_fullset_100223_annotated/all_chelicerate_proteins_annotated_plus_orthogroups_plus_clusters.csv\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "374a2d6b-8c0b-4af8-a2b2-441215207338",
   "metadata": {},
   "outputs": [],
   "source": [
    "#subsetting down to tick proteins (only have structural cluster information for tick proteins anway)\n",
    "tick_proteins = all_proteins_merged.loc[all_proteins_merged[\"is_tick\"] ==\"yes\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2ce47f7-ef25-4cea-b4b5-489de63fa2e5",
   "metadata": {},
   "source": [
    "### Getting counts and stats for different orthogroups are in each structural cluster to pick the orthogroups to move forward for trait mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a869ac6c-af73-4369-b908-460a41c6b96f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#HQ = high confidence PI clusters \n",
    "\n",
    "HQ_target_PIs = tick_proteins.loc[tick_proteins[\"PI_cluster\"].notna()]\n",
    "\n",
    "#LQ = low confidence PI clustes (sketchy clusters)\n",
    "LQ_target_PIs = tick_proteins.loc[tick_proteins[\"Low_Quality_PI_cluster\"].notna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3410b93e-51cb-432e-bb0e-3c6a6bbe9a3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ortho_counts(ortho_list, name_col):\n",
    "    #take a list of orthogroups and a column of their names (like PI cluster names) and count up how many proteins are in the OG\n",
    "    #also counts how many proteins are secreted, how many proteins were in carto, how many tick speices are in OG \n",
    "    \n",
    "    num_dict = defaultdict(list)\n",
    "\n",
    "    for orthogroup in ortho_list:\n",
    "        # proteins that were part of cartography run \n",
    "        carto = tick_proteins.loc[(tick_proteins[\"Orthogroup\"] == orthogroup) & (tick_proteins[\"LeidenCluster\"].notna())]\n",
    "        \n",
    "        # secreted carto proteins \n",
    "        carto_secreted = tick_proteins.loc[(tick_proteins[\"Orthogroup\"] == orthogroup) & (tick_proteins[\"deepsig_feature\"] == \"Signal peptide\") & (tick_proteins[\"LeidenCluster\"].notna())]\n",
    "\n",
    "        # all tick proteins\n",
    "        all_tick = tick_proteins.loc[(tick_proteins[\"Orthogroup\"] == orthogroup) ]\n",
    "\n",
    "        #secreted tick proteins\n",
    "        tick_secreted = tick_proteins.loc[(tick_proteins[\"Orthogroup\"] == orthogroup) & (tick_proteins[\"deepsig_feature\"] == \"Signal peptide\")]\n",
    "        #names associated with orthogroup \n",
    "        PI_names = set(tick_proteins.loc[tick_proteins[\"Orthogroup\"] == orthogroup][name_col])\n",
    "    \n",
    "        num_dict[\"Orthogroup\"].append(orthogroup)\n",
    "        num_dict[\"species_total\"].append(len(set(all_tick[\"Organism\"])))\n",
    "        num_dict[\"carto_total\"].append(len(carto))\n",
    "        num_dict[\"carto_secreted\"].append(len(carto_secreted))\n",
    "        num_dict[\"tick_total\"].append(len(all_tick))\n",
    "        num_dict[\"tick_secreted\"].append(len(tick_secreted))\n",
    "        num_dict[\"PI_cluster\"].append(PI_names)\n",
    "\n",
    "        #print(orthogroup, name_col, PI_names)\n",
    "        \n",
    "        \n",
    "\n",
    "    return(num_dict)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83d861c8-5974-4cdd-bbc2-c667d01c82cd",
   "metadata": {},
   "source": [
    "### Getting orthogroups and their sizes for the good high quality (HQ)  structural clusters "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f37c3087-a99a-455f-908e-8086b466214c",
   "metadata": {},
   "outputs": [],
   "source": [
    "HQ_PI_ortho = set(HQ_target_PIs.loc[HQ_target_PIs[\"Orthogroup\"].notna()][\"Orthogroup\"])\n",
    "HQ_ortho_counts = get_ortho_counts(HQ_PI_ortho,\"PI_cluster\" )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18c1cd86-57f2-42c3-bee0-565441d697c9",
   "metadata": {},
   "source": [
    "### Removing small orthogroups and/or orthogroups with a low number of secreted proteins\n",
    "This gives us 14 orthogroups of high confidence protease inhibitors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a88fd588-959a-446a-b3fc-547723e4bac8",
   "metadata": {},
   "outputs": [],
   "source": [
    "HQ_num_df = pd.DataFrame(HQ_ortho_counts)\n",
    "\n",
    "#adding some more calculations for filtering\n",
    "HQ_num_df[\"percent_in_carto\"] = 100*HQ_num_df[\"carto_total\"]/HQ_num_df[\"tick_total\"]\n",
    "HQ_num_df[\"percent_carto_secreted\"] = 100*HQ_num_df[\"carto_secreted\"]/HQ_num_df[\"carto_total\"]\n",
    "HQ_num_df[\"percent_tick_secreted\"] = 100*HQ_num_df[\"tick_secreted\"]/HQ_num_df[\"tick_total\"]\n",
    "\n",
    "#set some filtering criteria to get rid of really small orthogroups, as well as orthogroups without a lot of secreted proteins \n",
    "\n",
    "HQ_num_df_filtered = HQ_num_df.loc[HQ_num_df[\"tick_secreted\"] >=5]\n",
    "HQ_num_df_filtered = HQ_num_df_filtered.loc[HQ_num_df_filtered[\"percent_tick_secreted\"] >=5]\n",
    "HQ_num_df_filtered = HQ_num_df_filtered.loc[HQ_num_df_filtered[\"species_total\"]>=10]\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6286edf3-b0a5-40fa-8cb6-822572d289a4",
   "metadata": {},
   "source": [
    "### Getting orthogroups and their sizes for the low quality (LQ)  structural clusters "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f215c60e-5a34-455f-b317-9f04883d48d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "LQ_PI_ortho = set(LQ_target_PIs.loc[LQ_target_PIs[\"Orthogroup\"].notna()][\"Orthogroup\"])\n",
    "LQ_ortho_counts = get_ortho_counts(LQ_PI_ortho,\"Low_Quality_PI_cluster\" )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8db05a59-bdec-42c5-80ad-aa0d93f6ffa5",
   "metadata": {},
   "source": [
    "### Removing small orthogroups and/or orthogroups with a low number of secreted proteins. \n",
    "This gives us 22 orthogroups of low confidence protease inhibitors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d63ffcd8-1e2d-4d6c-a1fe-0d1926848f8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "LQ_num_df = pd.DataFrame(LQ_ortho_counts)\n",
    "LQ_num_df[\"percent_in_carto\"] = 100*LQ_num_df[\"carto_total\"]/LQ_num_df[\"tick_total\"]\n",
    "LQ_num_df[\"percent_carto_secreted\"] = 100*LQ_num_df[\"carto_secreted\"]/LQ_num_df[\"carto_total\"]\n",
    "LQ_num_df[\"percent_tick_secreted\"] = 100*LQ_num_df[\"tick_secreted\"]/LQ_num_df[\"tick_total\"]\n",
    "\n",
    "#set some filtering criteria to get rid of really small orthogroups, as well as orthogroups without a lot of secreted proteins \n",
    "#i used more stringent filtering criteria for these sketchy clusters compared to the good clusters \n",
    "\n",
    "LQ_num_df_filtered = LQ_num_df.loc[LQ_num_df[\"tick_secreted\"] >=10]\n",
    "LQ_num_df_filtered = LQ_num_df_filtered.loc[LQ_num_df_filtered[\"percent_tick_secreted\"] >=25]\n",
    "LQ_num_df_filtered = LQ_num_df_filtered.loc[LQ_num_df_filtered[\"percent_in_carto\"] >=50]\n",
    "\n",
    "LQ_num_df_filtered = LQ_num_df_filtered.loc[LQ_num_df_filtered[\"species_total\"] >=10]\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bef57626-ea50-4990-89f8-e701fac5d73a",
   "metadata": {},
   "source": [
    "### Writing out two seperate files for low and high confidence protease inhibitor orthogroups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4ef06574-8a01-4360-a47e-96de3a536088",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(HQ_num_df_filtered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8e795cf9-0812-40ea-a285-1fa70507d0bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orthogroup</th>\n",
       "      <th>species_total</th>\n",
       "      <th>carto_total</th>\n",
       "      <th>carto_secreted</th>\n",
       "      <th>tick_total</th>\n",
       "      <th>tick_secreted</th>\n",
       "      <th>PI_cluster</th>\n",
       "      <th>percent_in_carto</th>\n",
       "      <th>percent_carto_secreted</th>\n",
       "      <th>percent_tick_secreted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>OG0000039</td>\n",
       "      <td>15</td>\n",
       "      <td>408</td>\n",
       "      <td>202</td>\n",
       "      <td>424</td>\n",
       "      <td>204</td>\n",
       "      <td>{Kunitz-1, nan, Kunitz-2}</td>\n",
       "      <td>96.226415</td>\n",
       "      <td>49.509804</td>\n",
       "      <td>48.113208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>OG0000230</td>\n",
       "      <td>15</td>\n",
       "      <td>79</td>\n",
       "      <td>26</td>\n",
       "      <td>86</td>\n",
       "      <td>27</td>\n",
       "      <td>{CRISP-PIs, nan}</td>\n",
       "      <td>91.860465</td>\n",
       "      <td>32.911392</td>\n",
       "      <td>31.395349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>OG0008132</td>\n",
       "      <td>12</td>\n",
       "      <td>19</td>\n",
       "      <td>12</td>\n",
       "      <td>20</td>\n",
       "      <td>12</td>\n",
       "      <td>{Kunitz-1, nan, Kunitz-2}</td>\n",
       "      <td>95.000000</td>\n",
       "      <td>63.157895</td>\n",
       "      <td>60.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>OG0002510</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>42</td>\n",
       "      <td>12</td>\n",
       "      <td>{TIL, nan}</td>\n",
       "      <td>4.761905</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>28.571429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>OG0005278</td>\n",
       "      <td>10</td>\n",
       "      <td>36</td>\n",
       "      <td>33</td>\n",
       "      <td>39</td>\n",
       "      <td>33</td>\n",
       "      <td>{Cystatin, nan}</td>\n",
       "      <td>92.307692</td>\n",
       "      <td>91.666667</td>\n",
       "      <td>84.615385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>OG0006419</td>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>6</td>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "      <td>{Cystatin, nan}</td>\n",
       "      <td>84.615385</td>\n",
       "      <td>54.545455</td>\n",
       "      <td>46.153846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>OG0000017</td>\n",
       "      <td>15</td>\n",
       "      <td>476</td>\n",
       "      <td>164</td>\n",
       "      <td>478</td>\n",
       "      <td>164</td>\n",
       "      <td>{Serpin-2, Serpin-1, nan}</td>\n",
       "      <td>99.581590</td>\n",
       "      <td>34.453782</td>\n",
       "      <td>34.309623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>OG0000058</td>\n",
       "      <td>15</td>\n",
       "      <td>119</td>\n",
       "      <td>118</td>\n",
       "      <td>313</td>\n",
       "      <td>204</td>\n",
       "      <td>{TIL, nan}</td>\n",
       "      <td>38.019169</td>\n",
       "      <td>99.159664</td>\n",
       "      <td>65.175719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>OG0001121</td>\n",
       "      <td>14</td>\n",
       "      <td>21</td>\n",
       "      <td>5</td>\n",
       "      <td>22</td>\n",
       "      <td>5</td>\n",
       "      <td>{nan, Kunitz-2}</td>\n",
       "      <td>95.454545</td>\n",
       "      <td>23.809524</td>\n",
       "      <td>22.727273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>OG0000216</td>\n",
       "      <td>15</td>\n",
       "      <td>135</td>\n",
       "      <td>98</td>\n",
       "      <td>137</td>\n",
       "      <td>98</td>\n",
       "      <td>{Cystatin, nan}</td>\n",
       "      <td>98.540146</td>\n",
       "      <td>72.592593</td>\n",
       "      <td>71.532847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>OG0000221</td>\n",
       "      <td>15</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>160</td>\n",
       "      <td>44</td>\n",
       "      <td>{CRISP-PIs, nan}</td>\n",
       "      <td>3.750000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>27.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>OG0000155</td>\n",
       "      <td>15</td>\n",
       "      <td>186</td>\n",
       "      <td>106</td>\n",
       "      <td>188</td>\n",
       "      <td>106</td>\n",
       "      <td>{CRISP-PIs, nan}</td>\n",
       "      <td>98.936170</td>\n",
       "      <td>56.989247</td>\n",
       "      <td>56.382979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>OG0008075</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>{CRISP-PIs}</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>38.461538</td>\n",
       "      <td>38.461538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>OG0008448</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>10</td>\n",
       "      <td>15</td>\n",
       "      <td>10</td>\n",
       "      <td>{Serpin-2, Serpin-1}</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>66.666667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Orthogroup  species_total  carto_total  carto_secreted  tick_total  \\\n",
       "9    OG0000039             15          408             202         424   \n",
       "23   OG0000230             15           79              26          86   \n",
       "28   OG0008132             12           19              12          20   \n",
       "31   OG0002510             12            2               2          42   \n",
       "41   OG0005278             10           36              33          39   \n",
       "43   OG0006419             13           11               6          13   \n",
       "74   OG0000017             15          476             164         478   \n",
       "93   OG0000058             15          119             118         313   \n",
       "97   OG0001121             14           21               5          22   \n",
       "118  OG0000216             15          135              98         137   \n",
       "124  OG0000221             15            6               6         160   \n",
       "166  OG0000155             15          186             106         188   \n",
       "192  OG0008075             12           13               5          13   \n",
       "208  OG0008448             15           15              10          15   \n",
       "\n",
       "     tick_secreted                 PI_cluster  percent_in_carto  \\\n",
       "9              204  {Kunitz-1, nan, Kunitz-2}         96.226415   \n",
       "23              27           {CRISP-PIs, nan}         91.860465   \n",
       "28              12  {Kunitz-1, nan, Kunitz-2}         95.000000   \n",
       "31              12                 {TIL, nan}          4.761905   \n",
       "41              33            {Cystatin, nan}         92.307692   \n",
       "43               6            {Cystatin, nan}         84.615385   \n",
       "74             164  {Serpin-2, Serpin-1, nan}         99.581590   \n",
       "93             204                 {TIL, nan}         38.019169   \n",
       "97               5            {nan, Kunitz-2}         95.454545   \n",
       "118             98            {Cystatin, nan}         98.540146   \n",
       "124             44           {CRISP-PIs, nan}          3.750000   \n",
       "166            106           {CRISP-PIs, nan}         98.936170   \n",
       "192              5                {CRISP-PIs}        100.000000   \n",
       "208             10       {Serpin-2, Serpin-1}        100.000000   \n",
       "\n",
       "     percent_carto_secreted  percent_tick_secreted  \n",
       "9                 49.509804              48.113208  \n",
       "23                32.911392              31.395349  \n",
       "28                63.157895              60.000000  \n",
       "31               100.000000              28.571429  \n",
       "41                91.666667              84.615385  \n",
       "43                54.545455              46.153846  \n",
       "74                34.453782              34.309623  \n",
       "93                99.159664              65.175719  \n",
       "97                23.809524              22.727273  \n",
       "118               72.592593              71.532847  \n",
       "124              100.000000              27.500000  \n",
       "166               56.989247              56.382979  \n",
       "192               38.461538              38.461538  \n",
       "208               66.666667              66.666667  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HQ_num_df_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c4164ef9-b553-4be7-b20c-296a54486499",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "22"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(LQ_num_df_filtered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "20057803-a1ea-410b-8ff8-cac6e388112f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orthogroup</th>\n",
       "      <th>species_total</th>\n",
       "      <th>carto_total</th>\n",
       "      <th>carto_secreted</th>\n",
       "      <th>tick_total</th>\n",
       "      <th>tick_secreted</th>\n",
       "      <th>PI_cluster</th>\n",
       "      <th>percent_in_carto</th>\n",
       "      <th>percent_carto_secreted</th>\n",
       "      <th>percent_tick_secreted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>OG0000677</td>\n",
       "      <td>13</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>72</td>\n",
       "      <td>56</td>\n",
       "      <td>{BadTM-Kazal-IGF, nan}</td>\n",
       "      <td>69.444444</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>77.777778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>OG0007096</td>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>14</td>\n",
       "      <td>11</td>\n",
       "      <td>{BadTM-Kazal-IGF, nan}</td>\n",
       "      <td>78.571429</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>78.571429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>OG0009247</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>{BadTM-Kazal-A2M, nan}</td>\n",
       "      <td>84.615385</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>84.615385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>OG0009059</td>\n",
       "      <td>14</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>15</td>\n",
       "      <td>12</td>\n",
       "      <td>{WeakAnnotation-Kazal, nan}</td>\n",
       "      <td>80.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>80.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>OG0003200</td>\n",
       "      <td>14</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>48</td>\n",
       "      <td>37</td>\n",
       "      <td>{BadTM-Kazal-IGF, nan}</td>\n",
       "      <td>77.083333</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>77.083333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>OG0005859</td>\n",
       "      <td>15</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>30</td>\n",
       "      <td>16</td>\n",
       "      <td>{WeakAnnotation-Kazal, nan}</td>\n",
       "      <td>53.333333</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>53.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>OG0006622</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>{WeakAnnotation-Kazal, nan}</td>\n",
       "      <td>84.615385</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>84.615385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>OG0000480</td>\n",
       "      <td>14</td>\n",
       "      <td>33</td>\n",
       "      <td>5</td>\n",
       "      <td>52</td>\n",
       "      <td>18</td>\n",
       "      <td>{BadTM-Kazal-A2M, nan}</td>\n",
       "      <td>63.461538</td>\n",
       "      <td>15.151515</td>\n",
       "      <td>34.615385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>OG0001145</td>\n",
       "      <td>13</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>36</td>\n",
       "      <td>20</td>\n",
       "      <td>{WeakAnnotation-Kazal, nan}</td>\n",
       "      <td>55.555556</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>55.555556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>OG0001324</td>\n",
       "      <td>15</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>70</td>\n",
       "      <td>55</td>\n",
       "      <td>{BadTM-Kazal-IGF, nan}</td>\n",
       "      <td>78.571429</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>78.571429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>OG0001404</td>\n",
       "      <td>13</td>\n",
       "      <td>56</td>\n",
       "      <td>56</td>\n",
       "      <td>69</td>\n",
       "      <td>56</td>\n",
       "      <td>{BadTM-Kazal-IGF, nan}</td>\n",
       "      <td>81.159420</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>81.159420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>OG0007962</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>13</td>\n",
       "      <td>10</td>\n",
       "      <td>{BadTM-Kazal-IGF, nan}</td>\n",
       "      <td>76.923077</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>76.923077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>OG0002328</td>\n",
       "      <td>12</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>28</td>\n",
       "      <td>19</td>\n",
       "      <td>{BadTM-Kazal-IGF, nan}</td>\n",
       "      <td>67.857143</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>67.857143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>OG0002054</td>\n",
       "      <td>14</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>58</td>\n",
       "      <td>33</td>\n",
       "      <td>{BadTM-Kazal-IGF, nan}</td>\n",
       "      <td>56.896552</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>56.896552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>OG0009090</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>15</td>\n",
       "      <td>12</td>\n",
       "      <td>{BadTM-Kazal-IGF, BadTM-Kazal-A2M, nan}</td>\n",
       "      <td>80.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>80.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>OG0007868</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>{BadTM-Kazal-A2M, nan}</td>\n",
       "      <td>92.307692</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>92.307692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>OG0009632</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>{BadTM-Kazal-A2M, nan}</td>\n",
       "      <td>84.615385</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>84.615385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>OG0000890</td>\n",
       "      <td>14</td>\n",
       "      <td>76</td>\n",
       "      <td>63</td>\n",
       "      <td>81</td>\n",
       "      <td>63</td>\n",
       "      <td>{BadTM-Kazal-IGF, nan}</td>\n",
       "      <td>93.827160</td>\n",
       "      <td>82.894737</td>\n",
       "      <td>77.777778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>OG0007894</td>\n",
       "      <td>14</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>18</td>\n",
       "      <td>12</td>\n",
       "      <td>{BadTM-Kazal-A2M, nan}</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>66.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>OG0007881</td>\n",
       "      <td>15</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>18</td>\n",
       "      <td>16</td>\n",
       "      <td>{WeakAnnotation-Kazal, nan}</td>\n",
       "      <td>88.888889</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>88.888889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>OG0000204</td>\n",
       "      <td>15</td>\n",
       "      <td>42</td>\n",
       "      <td>6</td>\n",
       "      <td>78</td>\n",
       "      <td>35</td>\n",
       "      <td>{BadTM-Kazal-A2M, nan}</td>\n",
       "      <td>53.846154</td>\n",
       "      <td>14.285714</td>\n",
       "      <td>44.871795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>368</th>\n",
       "      <td>OG0001316</td>\n",
       "      <td>14</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>39</td>\n",
       "      <td>24</td>\n",
       "      <td>{BadTM-Kazal-A2M, nan}</td>\n",
       "      <td>61.538462</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>61.538462</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Orthogroup  species_total  carto_total  carto_secreted  tick_total  \\\n",
       "7    OG0000677             13           50              50          72   \n",
       "41   OG0007096             13           11              11          14   \n",
       "51   OG0009247             11           11              11          13   \n",
       "52   OG0009059             14           12              12          15   \n",
       "71   OG0003200             14           37              37          48   \n",
       "105  OG0005859             15           16              16          30   \n",
       "113  OG0006622             12           11              11          13   \n",
       "163  OG0000480             14           33               5          52   \n",
       "164  OG0001145             13           20              20          36   \n",
       "199  OG0001324             15           55              55          70   \n",
       "200  OG0001404             13           56              56          69   \n",
       "239  OG0007962             10           10              10          13   \n",
       "240  OG0002328             12           19              19          28   \n",
       "246  OG0002054             14           33              33          58   \n",
       "251  OG0009090             13           12              12          15   \n",
       "260  OG0007868             11           12              12          13   \n",
       "268  OG0009632             11           11              11          13   \n",
       "289  OG0000890             14           76              63          81   \n",
       "300  OG0007894             14           12              12          18   \n",
       "312  OG0007881             15           16              16          18   \n",
       "351  OG0000204             15           42               6          78   \n",
       "368  OG0001316             14           24              24          39   \n",
       "\n",
       "     tick_secreted                               PI_cluster  percent_in_carto  \\\n",
       "7               56                   {BadTM-Kazal-IGF, nan}         69.444444   \n",
       "41              11                   {BadTM-Kazal-IGF, nan}         78.571429   \n",
       "51              11                   {BadTM-Kazal-A2M, nan}         84.615385   \n",
       "52              12              {WeakAnnotation-Kazal, nan}         80.000000   \n",
       "71              37                   {BadTM-Kazal-IGF, nan}         77.083333   \n",
       "105             16              {WeakAnnotation-Kazal, nan}         53.333333   \n",
       "113             11              {WeakAnnotation-Kazal, nan}         84.615385   \n",
       "163             18                   {BadTM-Kazal-A2M, nan}         63.461538   \n",
       "164             20              {WeakAnnotation-Kazal, nan}         55.555556   \n",
       "199             55                   {BadTM-Kazal-IGF, nan}         78.571429   \n",
       "200             56                   {BadTM-Kazal-IGF, nan}         81.159420   \n",
       "239             10                   {BadTM-Kazal-IGF, nan}         76.923077   \n",
       "240             19                   {BadTM-Kazal-IGF, nan}         67.857143   \n",
       "246             33                   {BadTM-Kazal-IGF, nan}         56.896552   \n",
       "251             12  {BadTM-Kazal-IGF, BadTM-Kazal-A2M, nan}         80.000000   \n",
       "260             12                   {BadTM-Kazal-A2M, nan}         92.307692   \n",
       "268             11                   {BadTM-Kazal-A2M, nan}         84.615385   \n",
       "289             63                   {BadTM-Kazal-IGF, nan}         93.827160   \n",
       "300             12                   {BadTM-Kazal-A2M, nan}         66.666667   \n",
       "312             16              {WeakAnnotation-Kazal, nan}         88.888889   \n",
       "351             35                   {BadTM-Kazal-A2M, nan}         53.846154   \n",
       "368             24                   {BadTM-Kazal-A2M, nan}         61.538462   \n",
       "\n",
       "     percent_carto_secreted  percent_tick_secreted  \n",
       "7                100.000000              77.777778  \n",
       "41               100.000000              78.571429  \n",
       "51               100.000000              84.615385  \n",
       "52               100.000000              80.000000  \n",
       "71               100.000000              77.083333  \n",
       "105              100.000000              53.333333  \n",
       "113              100.000000              84.615385  \n",
       "163               15.151515              34.615385  \n",
       "164              100.000000              55.555556  \n",
       "199              100.000000              78.571429  \n",
       "200              100.000000              81.159420  \n",
       "239              100.000000              76.923077  \n",
       "240              100.000000              67.857143  \n",
       "246              100.000000              56.896552  \n",
       "251              100.000000              80.000000  \n",
       "260              100.000000              92.307692  \n",
       "268              100.000000              84.615385  \n",
       "289               82.894737              77.777778  \n",
       "300              100.000000              66.666667  \n",
       "312              100.000000              88.888889  \n",
       "351               14.285714              44.871795  \n",
       "368              100.000000              61.538462  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LQ_num_df_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "78ccc1c9-85b8-4dfd-8199-74ed6c06e15d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#writing out target orthogroups - handing off for trait mapping analysis \n",
    "HQ_num_df_filtered.to_csv(\"../datasheets/PI_orthogroups_high_qual_06112023.tsv\", sep = \"\\t\", index = False)\n",
    "LQ_num_df_filtered.to_csv(\"../datasheets/PI_orthogroups_low_qual_06112023.tsv\", sep = \"\\t\", index = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
